Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp

```diff
@@ -106,37 +122,73 @@ void visitSeedIdentifiers(const TrackProxy& track, Visitor visitor) {
   }
 }
 
-}  // namespace
-}  // namespace ActsExamples
+class BranchStopper {
+ public:
+  using Config =
+      std::optional<std::variant<Acts::TrackSelector::Config,
+                                 Acts::TrackSelector::EtaBinnedConfig>>;
 
-// Specialize std::hash for SeedIdentifier
-// This is required to use SeedIdentifier as a key in an `std::unordered_map`.
-template <class T, std::size_t N>
-struct std::hash<std::array<T, N>> {
-  std::size_t operator()(const std::array<T, N>& array) const {
-    std::hash<T> hasher;
-    std::size_t result = 0;
-    for (auto&& element : array) {
-      boost::hash_combine(result, hasher(element));
+  explicit BranchStopper(const Config& config) : m_config(config) {}
+
+  bool operator()(
+      const Acts::CombinatorialKalmanFilterTipState& tipState,
+      Acts::VectorMultiTrajectory::TrackStateProxy& trackState) const {
+    if (!m_config.has_value()) {
+      return false;
     }
-    return result;
+
+    bool haveCuts = std::visit(
+        [&](const auto& config) -> bool {
+          using T = std::decay_t<decltype(config)>;
+          if constexpr (std::is_same_v<T, Acts::TrackSelector::Config>) {
+            return true;
+          } else if constexpr (std::is_same_v<
+                                   T, Acts::TrackSelector::EtaBinnedConfig>) {
+            double theta = trackState.parameters()[Acts::eBoundTheta];
+            double eta = -std::log(std::tan(0.5 * theta));
+            return config.hasCuts(eta);
+          }
+        },
+        *m_config);
+
+    if (!haveCuts) {
+      return false;
+    }
+
+    const Acts::TrackSelector::Config& singleConfig = std::visit(
+        [&](const auto& config) -> const Acts::TrackSelector::Config& {
+          using T = std::decay_t<decltype(config)>;
+          if constexpr (std::is_same_v<T, Acts::TrackSelector::Config>) {
+            return config;
+          } else if constexpr (std::is_same_v<
+                                   T, Acts::TrackSelector::EtaBinnedConfig>) {
+            double theta = trackState.parameters()[Acts::eBoundTheta];
+            double eta = -std::log(std::tan(0.5 * theta));
+            return config.getCuts(eta);
+          }
+        },
+        *m_config);
+
+    if (tipState.nMeasurements >= singleConfig.minMeasurements) {
+      return false;
+    }
+
+    if (tipState.nHoles >= singleConfig.maxHoles) {
+      return true;
+    }
+
+    return false;
```

Comment:
alternatively, this could be rewritten like this:

```suggestion
    if (tipState.nHoles <= singleConfig.maxHoles) {
      return false;
    }

    if (tipState.nMeasurements < singleConfig.minMeasurements) {
      return true;
    }

    if (trackState.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
      return true;
    }

    return false;
```

Reply:
I am not sure about the last condition
```
    if (trackState.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
      return true;
    }
```
Why are we stopping in this case?

Reply:
This stops the track where you have more than two holes that isn't at the end of a track. In this case we can reject the track even if it has enough measurements.

I think the other logic I suggested above is probably clearer, but this one is simpler 😄 and fits your current design.

Reply:
I see - but then I am confused about the final `return false`.

In this case we have more holes than `maxHoles` and more measurements than `minMeasurements` but we keep the track?